### ISSUE TITLE
Visualize_topics randomness

### ISSUE THREAD

**Issue by Jason990407** [MaartenGr/BERTopic#2403]
### Have you searched existing issues?  🔎

- [x] I have searched and found no existing issues

### Desribe the bug

Thank you for your great package sir,

I just found a bug, it seems that even if I have set the random state in UMAP's parameter, the visualization of LDAvis's bubbles is totally random. The function .get_topic_info() is reproducible, but no matter what I try for the .visualize_topics(), it just does not work. I checked your _topics.py file, it looks like you also set the random state for the visualization in terms of UMAP, so I wonder whether the bubble chart could be reproducible, or I just need to save it in case the results vanished. 

### Reproduction

```python
from bertopic import BERTopic

```

### BERTopic Version

0.15.0

**Comment by MaartenGr** [comment]
This is quite an old version of BERTopic. Are you experiencing this issue with the latest version? I believe this was fixed a while ago.

**Comment by Jason990407** [comment]
> This is quite an old version of BERTopic. Are you experiencing this issue with the latest version? I believe this was fixed a while ago.

I have switched the latest version of BERTopic, and I tried it on the colab, the results looked all good, but once I tried the same version in local server, and I make sure every important package is the same as the ones I used on the colab, the results are still random

**Comment by Jason990407** [comment]
> This is quite an old version of BERTopic. Are you experiencing this issue with the latest version? I believe this was fixed a while ago.

I just tried another dataset on colab, unfortunately the figure is still cannot be reproducible, I wonder if it is because I also transfer the pre-saved embeddings to the model? My code looks like this:

with open('./data/留言板切词.txt', 'r', encoding='utf-8') as file:
  docs = file.readlines()
print('条数: ', len(docs))
print('预览第一条: ', docs[0])

embedding_model = SentenceTransformer('/home/jiaxing/.cache/huggingface/hub/models--aspire--acge_text_embedding/snapshots/652afc78466e53d1909c785c26f5b978021c4650',)
embeddings = np.load('./data/embedding_acge.npy')
print(embeddings.shape)

umap_model = UMAP(
  n_neighbors=50,
  n_components=5,
  min_dist=0.0,
  metric='cosine',
  random_state=42  
)

hdbscan_model = HDBSCAN(
  min_cluster_size=100,
  min_samples=100,
  metric='euclidean'
)

vectorizer_model = CountVectorizer()

topic_model = BERTopic(
  embedding_model=embedding_model,
  vectorizer_model=vectorizer_model,
  umap_model=umap_model,
  hdbscan_model=hdbscan_model,

topics, probs = topic_model.fit_transform(docs, embeddings= embeddings)

topic_model.get_topic_info()

topic_model.visualize_topics()

So basically I use another pipeline to pre-store the embeddings that are generated by the same model I use SentenceTransformer to load, and this code will just load the saved .npy file for saving time.

**Comment by MaartenGr** [comment]
Looking at the source code, if the **exact** same environment is being used, then `.visualize_topics` should be reproducible. If it isn't the same environment, then that might differ. I believe this might be a result of UMAP's specific implementation.

**Comment by makramab** [comment]
Hi! I am also encountering a similar issue with `visualize_topics()`.

<details>
<summary>Environment</summary>

- **BERTopic**: 0.17.3  
- **umap-learn**: 0.5.9.post2  

</details>

<details>
<summary>Minimal code</summary>

```python
from sentence_transformers import SentenceTransformer
from umap import UMAP
from hdbscan import HDBSCAN
from sklearn.feature_extraction.text import CountVectorizer
from bertopic.representation import KeyBERTInspired, MaximalMarginalRelevance, PartOfSpeech
from bertopic import BERTopic

# Embeddings
embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = embedding_model.encode(full_text_list, show_progress_bar=True)

# UMAP / HDBSCAN / Vectorizer
umap_model = UMAP(n_neighbors=15, n_components=5, min_dist=0.0,
                  metric="cosine", random_state=42)
hdbscan_model = HDBSCAN(min_cluster_size=150, metric="euclidean",
                        cluster_selection_method="eom", prediction_data=True)
vectorizer_model = CountVectorizer(stop_words="english", min_df=2, ngram_range=(1, 2))

# Representations
keybert_model = KeyBERTInspired()
pos_model = PartOfSpeech("en_core_web_sm")
mmr_model = MaximalMarginalRelevance(diversity=0.3)

representation_model = {
    "KeyBERT": keybert_model,
    "MMR": mmr_model,
    "POS": pos_model
}

# BERTopic
topic_model = BERTopic(
    embedding_model=embedding_model,
    umap_model=umap_model,
    hdbscan_model=hdbscan_model,
    vectorizer_model=vectorizer_model,
    representation_model=representation_model,
    top_n_words=10,
    verbose=True
)

# Train
topics, probs = topic_model.fit_transform(full_text_list, embeddings)

# Visualize
topic_model.visualize_topics()
```

</details>

Each time I call `visualize_topics()`, the bubble layout changes (the two pics below are coming from the same bertopic instance).

<img width="879" height="746" alt="Image" src="https://github.com/user-attachments/assets/c770fef4-9f1a-4a1b-ad12-57d1a1cf4b50" />


<img width="879" height="748" alt="Image" src="https://github.com/user-attachments/assets/ea8d3476-a326-435b-b389-18c4885eaad7" />

I tried testing things around and found out that in my case, the variability seems tied to the UMAP initialization used during plotting.  When initializing UMAP,  it defaults to [`init="spectral"`](https://github.com/lmcinnes/umap/blob/90871ffc135a0cf8e0866881f149b4d4cf562152/umap/umap_.py#L1477). With that, I get slightly different embeddings each run, even with a fixed `random_state`.  

Changing UMAP’s init to `init="random"` or `init="pca"` made the embeddings reproducible in my experiments.  

As `visualize_topics()` appears to initialize its own UMAP under the hood, this explains the randomness when calling `visualize_topics()` as the embeddings used for visuailzation is not deterministic.

https://github.com/MaartenGr/BERTopic/blob/144ab7b63a6ae724972562cb62e73784c179c118/bertopic/plotting/_topics.py#L102

Is there a strict reason why `visualize_topics()` needs to use `init="spectral"`?  
If not, would it make sense to either:
- Reuse the configured `umap_model` from the pipeline, or  
- Expose an argument (e.g. `umap_init=...` or `umap_model=...`) so we can choose `init="random"` or `init="pca"` for reproducible layouts?

Happy to test a PR or share more diagnostics if that helps. Thanks!

**Comment by Jason990407** [comment]
> Looking at the source code, if the **exact** same environment is being used, then `.visualize_topics` should be reproducible. If it isn't the same environment, then that might differ. I believe this might be a result of UMAP's specific implementation.

I try to use the Kmeans instead of the HDBSCAN, the bubble charts however can be reproducible, I still dont know the mechanism behind it.

**Comment by makramab** [comment]
> I try to use the Kmeans instead of the HDBSCAN, the bubble charts however can be reproducible, I still dont know the mechanism behind it.

I tried replicating this behavior with `KMeans` instead of `HDBSCAN` and noticed something interesting:

```python
from sklearn.cluster import KMeans

cluster_model = KMeans(n_clusters=50)   # Deterministic result in my runs
cluster_model = KMeans(n_clusters=150)  # Non-deterministic result
```

When I use `n_clusters=50`, the layout from `visualize_topics()` becomes reproducible.  
But when I increase to `n_clusters=150`, it goes back to being non-deterministic.

This lines up with what was explained earlier about `init="spectral"` in UMAP (sorry but this is going to get a little bit technical into UMAP inner workings),

For a given initial embedding, spectral init can sometimes yield deterministic eigenvectors And other times it can drift into different (but equally valid) bases, depending on eigenvalue gaps and graph geometry.  That means depending on the setup (like number of clusters or dataset size), we might *luckily* land on a configuration that results in an embedding that is deterministic for UMAP with init="spectral".

So in your case, I believe your embedding setup happened to be one of the “deterministic” ones for `spectral`, which is why you observed reproducibility. But it won’t hold in general.

I’m attaching my embeddings here in case you want to test on your side. 

[topic_embeddings.zip](https://github.com/user-attachments/files/21967047/topic_embeddings.zip)

You may want to also try changing `n_clusters` and see whether it flips between deterministic and non-deterministic layouts for you as well.

**Comment by makramab** [comment]
Hi, sorry for following up again.

I created two Google Colab notebooks to illustrate the issue more clearly by just running the example code from the [Best Practice](https://maartengr.github.io/BERTopic/getting_started/best_practices/best_practices.html) section in the documentation.

- **First notebook**: [Colab link](https://colab.research.google.com/drive/1hYtT7y_Um5elqtj9ebpqflXqtE2Yu_op?authuser=1#scrollTo=lu-nINKjlyKN)  
  Using the current latest BERTopic release. Each time I call `visualize_topics()`, it produces a different visualization.  

![Image](https://github.com/user-attachments/assets/eadfaa48-d04c-4dd4-9098-8e5711bf7ef1)

- **Second notebook**: [Colab link](https://colab.research.google.com/drive/1H9Fkg_ZYZmpEEKso6Z6XL5q5s0NwfcsY?authuser=1#scrollTo=_f0pL234qpzC)  
  This one uses a forked version where I explicitly set `init="random"` in the UMAP step inside `visualize_topics()`.  
  With this change, the visualization becomes reproducible across runs.  

![Image](https://github.com/user-attachments/assets/efd22760-f88e-41f1-887e-ddd6d30936ef)

---

I also opened a PR for this change.

At this stage, I’m not entirely sure about the trade-offs of explicitly using `init="random"` instead of using the default `init="spectral"`. It might need further discussion to evaluate whether this should be the default behavior or if there's other more preferable alternatives.
